### ISSUE TITLE
[MPS] BatchNorm2d backward produces wrong gradients on channels_last/non-contiguous (weight.grad 7 OOM off)

### ISSUE THREAD

**Issue by npinto** [pytorch/pytorch#175189]
## Summary

`nn.BatchNorm2d` backward on MPS produces **wrong gradients** when the input is `channels_last` or non-contiguous (via `permute`). Forward is correct. The backward bug affects:
- `weight.grad`: 7 orders of magnitude wrong (CPU: ~1e-7, MPS: ~13)
- `input.grad`: 3-7x wrong
- `bias.grad`: always correct

Broken in **ALL modes**: training, eval, `affine=False`, and `F.batch_norm`.

**LSTM backward** also produces wrong input gradients (~0.5-0.7 off) on non-contiguous inputs. Weight grads are fine.

**NOT affected:** `InstanceNorm2d`, `GroupNorm`, `LayerNorm`, `RNN`, `GRU` — all handle channels_last backward correctly.

**Impact:** Training any model with BatchNorm + channels_last on MPS diverges within a few steps. Loss reaches 10^13 within 7 training steps due to cascading wrong weight updates.

**Workaround:** `.contiguous()` before BatchNorm.

## Reproducer (BatchNorm2d backward)

```python
import torch
import torch.nn as nn
assert torch.backends.mps.is_available()
torch.manual_seed(42)

bn_cpu = nn.BatchNorm2d(3)
bn_mps = nn.BatchNorm2d(3).to("mps")
bn_mps.load_state_dict(bn_cpu.state_dict())

base = torch.randn(4, 3, 8, 8)
cpu_x = base.clone().to(memory_format=torch.channels_last).requires_grad_(True)
mps_x = base.to("mps").clone().to(memory_format=torch.channels_last).requires_grad_(True)

bn_cpu(cpu_x).sum().backward()
bn_mps(mps_x).sum().backward()
torch.mps.synchronize()

ig = (cpu_x.grad - mps_x.grad.cpu()).abs().max().item()
wg = (bn_cpu.weight.grad - bn_mps.weight.grad.cpu()).abs().max().item()
bg = (bn_cpu.bias.grad - bn_mps.bias.grad.cpu()).abs().max().item()
print(f"input.grad diff: {ig:.4f}")   # ~3-7x wrong
print(f"weight.grad diff: {wg:.4f}")  # ~13 (7 OOM wrong)
print(f"bias.grad diff: {bg:.6f}")    # ~0 (correct)
assert wg < 0.01, f"weight.grad wrong: diff={wg}"
# FAILS: weight.grad is ~13 instead of ~1e-7
```

## Reproducer (training divergence)

```python
import torch
import torch.nn as nn
assert torch.backends.mps.is_available()

model = nn.Sequential(nn.Conv2d(3, 16, 3, padding=1), nn.BatchNorm2d(16), nn.ReLU())
model.to("mps").to(memory_format=torch.channels_last)
opt = torch.optim.SGD(model.parameters(), lr=0.01)

for step in range(10):
    x = torch.randn(4, 3, 32, 32, device="mps").to(memory_format=torch.channels_last)
    loss = model(x).sum()
    opt.zero_grad()
    loss.backward()
    opt.step()
    print(f"Step {step}: loss={loss.item():.2f}")
# Loss diverges rapidly: step 0 ~normal, step 3 ~62, step 7 ~47 trillion
```

Related: #174345 (forward-only offset bug, different root cause)

## Environment

- PyTorch 2.10.0
- macOS 15.4 (Sequoia)
- Apple M2 Max
- Python 3.13

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @bobrenjc93 @jamesr66a @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk

**Comment by qflen** [comment]
Fix landed in 12a9ea264bf805a66cd87e19e767ab23c2f59fef (#188371), same approach as my PR that I closed as redundant.

This can be closed @malfet.

**Comment by malfet** [comment]
Thank you for the ping, closed.
